drivers: can_sam0: added clock configuration for SAME5x devices

sebastianschlupp opened the pull request:
drivers: can_sam0: added clock configuration for SAME5x devices

also modified the following to support CAN:
soc: same51 and same54: added DFLL48 frequency information
dts: arm: atmel: added can peripherals to same5x.dtsi

Signed-off-by: Sebastian Schlupp <sebastian.schlupp@gmail.com>

sebastianschlupp committed:
soc: same51 and same54: added DFLL48 frequency information

Specified the value 48000000 for the DFLL48 clock source

Signed-off-by: Sebastian Schlupp <sebastian.schlupp@gmail.com>

sebastianschlupp commented:
Removed the trailing spaces in the same5x.dtsi file as requested.
Sorry for the mess with merging / reverting, etc. Hope everything is fine now.

henrikbrixandersen commented on drivers/can/can_sam0.c:
```diff
@@ -94,19 +95,30 @@ static int can_sam0_get_core_clock(const struct device *dev, uint32_t *rate)
-
-	*rate = SOC_ATMEL_SAM0_OSC48M_FREQ_HZ / (sam_cfg->divider);
-
+	#if (CONFIG_SOC_SERIES_SAME51 || CONFIG_SOC_SERIES_SAME54)
```
```suggestion
#if defined(CONFIG_SOC_SERIES_SAME51) || defined(CONFIG_SOC_SERIES_SAME54)
```

henrikbrixandersen commented on drivers/can/can_sam0.c:
```diff
@@ -94,19 +95,30 @@ static int can_sam0_get_core_clock(const struct device *dev, uint32_t *rate)
+	#if (CONFIG_SOC_SERIES_SAME51 || CONFIG_SOC_SERIES_SAME54)
+	/*DFFL has to be used as clock source for the ATSAME51/54 family of SoCs*/
+		*rate = SOC_ATMEL_SAM0_DFLL48_FREQ_HZ / (sam_cfg->divider);
+	#elif (CONFIG_SOC_SERIES_SAMC21)
```
```suggestion
#elif defined(CONFIG_SOC_SERIES_SAMC21)
```

henrikbrixandersen commented on drivers/can/can_sam0.c:
```diff
@@ -94,19 +95,30 @@ static int can_sam0_get_core_clock(const struct device *dev, uint32_t *rate)
+	#elif (CONFIG_SOC_SERIES_SAMC21)
+	/*OSC48M has to be used as clock source for the ATSAMC21 family of SoCs*/
+		*rate = SOC_ATMEL_SAM0_OSC48M_FREQ_HZ / (sam_cfg->divider);
+	#endif
```
```suggestion
#endif
```

henrikbrixandersen commented on soc/arm/atmel_sam0/same51/soc.h:
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2019 ML!PA Consulting GmbH
+ * Copyright (c) 2023 Sebastian Schlupp
```
Normally, adding a few lines to an existing file does not warrant a copyright notice. Otherwise our files would have _really_ many copyright lines.

henrikbrixandersen reviewed (changes requested):
Please fix indentation and if-clauses as suggested for all occurrences.

sebastianschlupp commented:
I did make a mess with the 3 commits which should have been 1. Would it be best to just close the pull-request and open a new one on a new branch with a single commit or should I attempt to get rid of the last commits with the "git reset --hard HEAD~2" command and just amend the required changes to the first commit? I feel like it's the moment to stop attempting to fix it on my own while digging myself deeper and deeper and to wait for feedback of which direction to go :P

sebastianschlupp commented:
Considering the failures on the nucleo_f429zi which are completely unrelated to this is there anything else I'd need to do? First contribution here so not exactly if I just need to wait or if something needs to be done by me.

nandojve commented:
Hi @sebastianschlupp ,

You should split this PR in 3 commit:

1- dts: arm: atmel: added can peripherals to same5x.dtsi
2- soc: same51 and same54: added DFLL48 frequency information
3- update CAN driver

These 3 commit should be rebased on top of current main.

sebastianschlupp committed:
dts: arm: atmel: same5x: added CAN(0 and 1) peripheral description

Added default parameters for CAN peripherals according to the datasheet.

Signed-off-by: Sebastian Schlupp <sebastian.schlupp@gmail.com>

sebastianschlupp committed:
drivers: can_sam0: added clock configuration for SAME5x devices

Added clock source configuration depending on SAM SoC series.

Signed-off-by: Sebastian Schlupp <sebastian.schlupp@gmail.com>

sebastianschlupp commented:
> Hi @sebastianschlupp ,
> 
> You should split this PR in 3 commit:
> 
> 1- dts: arm: atmel: added can peripherals to same5x.dtsi 2- soc: same51 and same54: added DFLL48 frequency information 3- update CAN driver
> 
> These 3 commit should be rebased on top of current main.

done as requested

str4t0m reviewed (changes requested):
Please fix the commit messages and indentation of the code as reported by the compliance checks (No leading spaces, ... ).

sebastianschlupp commented:
> Please fix the commit messages and indentation of the code as reported by the compliance checks (No leading spaces, ... ).

done as requested

henrikbrixandersen commented:
> done as requested

Unfortunately, you still missed some. Please take a look at the CI job output (available here for your current push on this PR: https://github.com/zephyrproject-rtos/zephyr/actions/runs/7058482799/job/19221219219?pr=65513) and fix any issues reported there.

henrikbrixandersen commented:
@str4t0m Please revisit.

str4t0m reviewed (changes requested):
Thanks for addressing all the issues!
I didn't check the rationale of the clock source selection in the datasheet, but everything looks fine for me now.
Just one NIT regarding the interrupt priority definitions.

Also, I assume you have tested this on a custom board, but don't have access to the atsame54_xpro board? Because if you would have access to the board, it would be nice to also add definitions for the board such that the driver can be built in CI for this SoC and any regression is detected earlier.

sebastianschlupp commented:
> Thanks for addressing all the issues! I didn't check the rationale of the clock source selection in the datasheet, but everything looks fine for me now. Just one NIT regarding the interrupt priority definitions.
> 
> Also, I assume you have tested this on a custom board, but don't have access to the atsame54_xpro board? Because if you would have access to the board, it would be nice to also add definitions for the board such that the driver can be built in CI for this SoC and any regression is detected earlier.

Regarding the rationale for clock source selection, the E5x devices don't have the internal 48MHz oscillator which is used by default for the SAMC21. They have the possibility to use either an external oscillator (XOSC0, XOSC1, XOSC32K) directly or the output of the digital frequency locked loop (fixed to 48MHz). Because this is enabled by default, also fixed in value (like the internal 48MHz oscillator of the C21) and because it has the same value the driver was designed to work with, I chose that as a source. The source for the DFLL48M can be either the internal 32kHz oscillator or an external 32kHz oscillator. In both of the atsame54_xpro boards as well as the atsame51_cnano boards an external 32kHz oscillator is available and used.

Regarding the interrupts, I have only used LINE_0 interrupt initially. After asking about this on discord I have been pointed to the question raised about exactly this topic here: https://github.com/zephyrproject-rtos/zephyr/discussions/61014
I have compared by changes with the ones mentioned by the OP and decided to add the LINE_1 interrupt as well for possible other, future uses. I didn't really spend time on deciding which priority to select. Want me to change this as follows?
interrupts = <78 0>, <78 0>; for both CAN channels?

sebastianschlupp commented on dts/arm/atmel/same5x.dtsi:
```diff
@@ -25,5 +26,33 @@
+		can0: can@42000000 {
+			compatible = "atmel,sam0-can";
+			reg = <0x42000000 0x400>;
+			interrupts = <78 0>, <78 1>;
```
@str4t0m please see reply in thread

str4t0m commented:
> Regarding the interrupts, I have only used LINE_0 interrupt initially. After asking about this on discord I have been pointed to the question raised about exactly this topic here: #61014
> I have compared by changes with the ones mentioned by the OP and decided to add the LINE_1 interrupt as well for possible other, future uses. I didn't really spend time on deciding which priority to select. Want me to change this as follows?
> interrupts = <78 0>, <78 0>; for both CAN channels?

Thanks for the explanations. Yeah that is how I had the change in mind.
